housekeeping/src/main/groovy/whelk/housekeeping/NotificationSender.groovy

```diff
@@ -0,0 +1,429 @@
+package whelk.housekeeping
+
+import org.simplejavamail.api.email.Email
+import org.simplejavamail.api.mailer.Mailer
+import org.simplejavamail.email.EmailBuilder
+import org.simplejavamail.mailer.MailerBuilder
+import whelk.Document
+import whelk.JsonLd
+import whelk.Whelk
+import groovy.transform.CompileStatic
+import groovy.util.logging.Log4j2 as Log
+import whelk.util.PropertyLoader
+
+import java.sql.Connection
+import java.sql.PreparedStatement
+import java.sql.ResultSet
+import java.sql.Timestamp
+import java.time.Instant
+import java.time.ZoneOffset
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+@CompileStatic
+@Log
+class NotificationSender extends HouseKeeper {
+
+    private final String STATE_KEY = "Email notifications"
+    private String status = "OK"
+    private final Whelk whelk
+    private final Mailer mailer
+    private final String senderAddress
+
+    public NotificationSender(Whelk whelk) {
+        this.whelk = whelk
+        Properties props = PropertyLoader.loadProperties("secret")
+        if (props.containsKey("smtpServer") &&
+                props.containsKey("smtpPort") &&
+                props.containsKey("smtpSender") &&
+                props.containsKey("smtpUser") &&
+                props.containsKey("smtpPassword"))
+            mailer = MailerBuilder
+                .withSMTPServer(
+                        (String) props.get("smtpServer"),
+                        Integer.parseInt((String)props.get("smtpPort")),
+                        (String) props.get("smtpUser"),
+                        (String) props.get("smtpPassword")
+                ).buildMailer()
+        senderAddress = props.get("smtpSender")
+    }
+
+    public String getName() {
+        return "Notifications sender"
+    }
+
+    public String getStatusDescription() {
+        return status
+    }
+
+    public void trigger() {
+        // Build a multi-map of library -> list of settings objects for that library's users
+        Map<String, List<Map>> heldByToUserSettings = new HashMap<>();
+        {
+            List<Map> allUserSettingStrings = whelk.getStorage().getAllUserData()
+            for (Map settings : allUserSettingStrings) {
+                if (!settings["notificationEmail"])
+                    continue
+                settings?.requestedNotifications.each { request ->
+                    if (!request instanceof Map)
+                        return
+                    if (!request["heldBy"])
+                        return
+
+                    String heldBy = request["heldBy"]
+                    if (!heldByToUserSettings.containsKey(heldBy))
+                        heldByToUserSettings.put(heldBy, [])
+                    List userSettingsForThisHeldBy = heldByToUserSettings[heldBy]
+                    userSettingsForThisHeldBy.add(settings)
```

Comment:
I think the existing code _is_ wrong? Otherwise what's the point of
```
List userSettingsForThisHeldBy = heldByToUserSettings[heldBy]
userSettingsForThisHeldBy.add(settings)
```
? `userSettingsForThisHeldBy` is not used anywhere else in the code.

Reply:
You can (as I did) get a reference to a list, and modify that list, without invalidating other references to that list (its changed in place).
The existing code, and your suggestion do the exact same thing, only yours is shorter and more readable.

Or are we talking about different things?

Reply:
No, you're right. Sorry, my brain is still a bit fried. Nevermind :D

Reply:
:+1: